File: Python/Permutations II.py
# Given a collection of numbers that might contain duplicates, return all possible unique permutations.

# For example,
# [1,1,2] have the following unique permutations:
# [1,1,2], [1,2,1], and [2,1,1].


class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
	def permuteUnique(self, num):
		result = []
		visited = [0 for i in range(0,len(num))]
		solution=[]
		step = 0
		num.sort()
		if len(num) == 0:
			return []
		self.findpermute(visited,solution,result,num,step)
		return result
	def findpermute(self,visited,solution,result,num,step):
		if step == len(num):
			result.append(list(solution))
			return
		for i in range(0,len(num)):
			if visited[i] == 0:
				if i > 0 and num[i] == num[i-1] and visited[i-1] == 0:
					continue
				visited[i] = 1
				solution.append(num[i])
				self.findpermute(visited,solution,result,num,step+1)
				solution.pop()
				visited[i]=0

File: Java/Permutations II.java
class Permutations II{
	public ArrayList<ArrayList<Integer>> permuteUnique(int[] num) {
        ArrayList<Integer> sol = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(num == null || num.length == 0)
        	return res;
        int[] visited = new int[num.length];
        Arrays.sort(num);
        genPermute(num,visited,sol,res);
        return res;
    }
    private void genPermute(int[] num,int[] visited, ArrayList<Integer> sol, ArrayList<ArrayList<Integer>> res){
    	if(sol.size() == num.length){
    		res.add(new ArrayList<Integer>(sol));
    		return;
    	}
    	for(int i=0;i<num.length;i++){
    		if(visited[i] == 0){
    			if(i>0 && num[i] == num[i-1] && visited[i-1] == 0)
    				continue;
    			visited[i] = 1;
    			sol.add(num[i]);
    			genPermute(num,visited,sol,res);
    			sol.remove(sol.size()-1);
    			visited[i] = 0;
    		}
    	}
    }
} 